### PR TITLE
#161254667 Like and dislike  an article

### PR DESCRIPTION
**Pull request by Gidraf**
#### What does this PR do?
```Like and dislike an article```
#### Description of Task to be completed?
```This PR makes users to appreciate those authors who created good articles by liking the article it also makes user to encourage those who didn't meet the expectation by disliking ```

#### How should this be manually tested?
- Clone this repo.
``` 
$ git clone https://github.com/andela/ah-orcas.git
```
- Navigate to this branch.
``` 
$ git clone git checkout  ft-like-dislike-article-161254667
```
- create a python virtual environment in your local machine and start it.
```code
$ virtualenv env
$ source env/bin/activate
```
- Create a .env file and run it. Check sample.env file to configure yours.
```
$ source .env
```
- Run the app
```
$ python manage.py runserver
```
- Test the endpoint.
```
api/users/article/dislike/<str:slug>/
api/users/article/like/<str:pk>/
```
#### What are the relevant pivotal tracker stories?
 [#161254667](https://www.pivotaltracker.com/story/show/161254667)

### CODE REVIEW

File: authors/apps/article/views.py [hunk 1]
```diff
@@ -142,8 +155,125 @@ def post(self, request, **kwargs):
...
+            return Response(
+                {"response": disliked_article_response},
+                status=status.HTTP_200_OK)
+        except Exception as e:
+            LikeDislikeArticle(
+                liker=disliker,
+                article=disliked_article,
+                is_disliked=True).save()
+            return Response(
+                {"response": "article successfully disliked"})
```
Comment (code-sleuth) [inline authors/apps/article/views.py]: The limit is 250 lines of code per file, why do we have 279 here?

File: authors/apps/article/tests/test_like_an_article.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                "slug": self.slug,
+                "rate": 3
+            }
+        }
+        self.data = {}
+        self.like_url = os.environ["URL"] + \
+            "api/article/" + "like/" + self.slug + "/"
+        self.view_likes_url = os.environ["URL"] + "api/article/rate/"
+        self.articles_url = os.environ["URL"] + "api/article/"
+        self.create_articles_url = os.environ["URL"] + "api/article/create"
```
Comment (Eyansky) [inline authors/apps/article/tests/test_like_an_article.py]: For modularity, instead of repeating the setup for every test. Why dont you create a base setup file so that they could share the setup files. It will also help anyone else writing tests could just add up to that.

File: authors/apps/article/views.py [hunk 3]
```diff
@@ -82,11 +95,11 @@ def perform_update(self, serializer):
         serializer.save(user=self.request.user)
 
 
-class ArticleRate(CreateAPIView):
+class ArticleRate(APIView):
```
Comment (Eyansky) [inline authors/apps/article/views.py]: Why did you change this to APIView?

File: authors/apps/article/views.py (hunk 1 above)
Comment (Gidraf) [reply authors/apps/article/views.py]: The reason I haven't included the http status code is because of the 250 of limit line of code @evamaina 
@code-sleuth the line of code is 250 but the blank space makes it 279

File: authors/apps/article/views.py (hunk 3 above)
Comment (Gidraf) [reply authors/apps/article/views.py]: @Eyansky Because is a get method not post

File: authors/apps/article/tests/test_dislike_an_article.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            "api/article/" + "dislike/" + self.slug + "/"
+        self.client.credentials(HTTP_AUTHORIZATION='Bearer ' + self.token)
+        self.client.post(self.dislike_url, self.data, format='json')
+        response = self.client.post(self.dislike_url, self.data, format='json')
+        self.assertIn("article successfully undisliked", str(response.data))
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_get_dislikes_article(self):
+        """
+        test whether quality
```
Comment (TheSteelGuy) [inline authors/apps/article/tests/test_dislike_an_article.py]: test whether quality is a wrong doc string

File: authors/apps/article/tests/test_like_an_article.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import os
+from ...authentication.models import User
+from ..models import Article
+from rest_framework.test import APIClient
+from rest_framework.test import APITestCase
+from rest_framework import status
+
+
+class TestLikeArticle(APITestCase):
+    """ class for testing email verification"""
```
Comment (TheSteelGuy) [inline authors/apps/article/tests/test_like_an_article.py]: wrong doc string here..

File: authors/apps/article/tests/test_like_an_article.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        """
+        Test whether like request without token will fail
+        """
+        response = self.client.post(self.like_url, self.data, format='json')
+        self.assertIn(
+            "Authentication credentials were not provided", str(
+                response.data))
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_like_article_of_article_not_found(self):
```
Comment (TheSteelGuy) [inline authors/apps/article/tests/test_like_an_article.py]: wrong doc string

File: authors/apps/article/tests/test_like_an_article.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        self.like_url = os.environ["URL"] + \
+            "api/article/" + "like/" + slug + "/"
+        self.client.credentials(HTTP_AUTHORIZATION='Bearer ' + self.token)
+        response = self.client.post(self.like_url, self.data, format='json')
+        self.assertIn("No article found", str(response.data))
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+
+    def test_like_article(self):
+        """
+        test whether like request with
```
Comment (TheSteelGuy) [inline authors/apps/article/tests/test_like_an_article.py]: wrong doc string

File: authors/apps/article/tests/test_like_an_article.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        self.like_url = os.environ["URL"] + \
+            "api/article/" + "like/" + self.slug + "/"
+        self.client.credentials(HTTP_AUTHORIZATION='Bearer ' + self.token)
+        response = self.client.post(self.like_url, self.data, format='json')
+        self.assertIn("article successfully liked", str(response.data))
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_like_twice_article(self):
+        """
+        test whether like request to an
```
Comment (TheSteelGuy) [inline authors/apps/article/tests/test_like_an_article.py]: wrong doc string

File: authors/apps/article/tests/test_like_an_article.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            "api/article/" + "like/" + self.slug + "/"
+        self.client.credentials(HTTP_AUTHORIZATION='Bearer ' + self.token)
+        self.client.post(self.like_url, self.data, format='json')
+        response = self.client.post(self.like_url, self.data, format='json')
+        self.assertIn("article successfully unliked", str(response.data))
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_get_likes_article_with_no_likes(self):
+        """
+        test whether quality
```
Comment (TheSteelGuy) [inline authors/apps/article/tests/test_like_an_article.py]: wrong doc string

File: authors/apps/article/tests/test_like_an_article.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+        self.like_url = os.environ["URL"] + \
+            "api/article/" + "like/" + self.slug + "/"
+        self.client.credentials(HTTP_AUTHORIZATION='Bearer ' + self.token)
+        response = self.client.post(self.like_url, self.data, format='json')
+        response = self.client.get(self.like_url, format='json')
+        self.assertIn("1", str(response.data))
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_get_like_article_of_article_not_found(self):
```
Comment (TheSteelGuy) [inline authors/apps/article/tests/test_like_an_article.py]: wrong doc string

File: authors/apps/article/views.py [hunk 11]
```diff
@@ -142,8 +155,125 @@ def post(self, request, **kwargs):
...
+        try:
+            disliked_article = get_article(kwargs)
+        except Exception:
+            return Response({"response": "No article found"},
+                            status=status.HTTP_204_NO_CONTENT)
+        dislikes = LikeDislikeArticle.objects.filter(
+            article=disliked_article)
+        dislikes_number = [
+            dislike for dislike in dislikes if dislike.is_disliked]
+        return Response({"response": len(dislikes_number)})
```
Comment (TheSteelGuy) [inline authors/apps/article/views.py]: what is the status code here

File: authors/apps/article/views.py [hunk 12]
```diff
@@ -142,8 +155,125 @@ def post(self, request, **kwargs):
...
+                dislike.save()
+                return Response(
+                    {"response": undisliked_response})
+            dislike.is_liked = False
+            dislike.is_disliked = True
+            dislike.save()
+            return Response(
+                {"response": disliked_article_response},
+                status=status.HTTP_200_OK)
+        except Exception as e:
```
Comment (TheSteelGuy) [inline authors/apps/article/views.py]: there is no need for saying Exception as e if you are not using the e somewhere. This is supposed to shorten your work which is not the case here so its not needed.
